### ISSUE TITLE
`flask.g` not reset when using  `test_client()` inside app context

### ISSUE THREAD

**Issue by malthejorgensen** [pallets/flask#2567]
Using the `test_client()` inside an app context, seemingly prevents the `flask.g`-object from being reset.

A test case like this 

```python
app = flask.Flask(__name__)

@app.route('/')
def index():
    if hasattr(flask.g, 'value'):
        flask.g.value += 1
    else:
        flask.g.value = 0
    return bytes(flask.g.value)

with app.app_context():
    with app.test_client() as client:
        resp = client.get('/')
        # `resp.data` will be '0'

        resp = client.get('/')
        # `resp.data` will be '1' here
```

will return `'1'` from the second request made by the test client. As the `flask.g`-object is local to the request, I would expect a `'0'` to be returned.

This may outside of the scope of supported use-cases, or simply the way it _should_ behave. I can accept that, but I just wanted it here at least for future users.

### Rationale for using `test_client()` inside an app context

The reason I do this is to be able to use `url_for()` inside my view tests, and have code like 

```python
response = client.post(
    url_for('blueprint_api.add_user'),
    json={'name': 'John', 'email': 'john@example.org'},
)
```

Basically I believe that my view tests shouldn't be testing a hard-coded URL (the single source of truth for that should be the route), but rather the logic of the view function.

### Environment

* Python version: 2.7.13
* Flask version: 0.13.0
* Werkzeug version: 0.13

**Comment by timster** [comment]
You're creating an application context, and `flask.g` persists through the application context.

If you don't wrap in `app.app_context()` it works properly.

```python
with app.test_client() as client:
    resp = client.get('/')
    # `resp.data` will be '0'

    resp = client.get('/')
    # `resp.data` will still be '0' here
```

**Comment by wgwz** [comment]
http://flask.pocoo.org/docs/dev/api/#flask.g

>Starting with Flask 0.10 this is stored on the application context and no longer on the request context which means it becomes available if only the application context is bound and not yet a request. This is especially useful when combined with the Faking Resources and Context pattern for testing.

This was interesting for me to look into! But it looks like you're misunderstanding where `flask.g` is stored. It looks like this is the intended behavior. It is perfectly reasonable to want to use `url_for` tests. 

This is the MCVE that worked for me: 
```import flask                                                                   
                                                                               
app = flask.Flask(__name__)                                                    
app.config.update({'SERVER_NAME':'test'})                                      
                                                                               
@app.route('/')                                                                
def index():                                                                   
    if hasattr(flask.g, 'value'):                                              
        flask.g.value += 1                                                     
    else:                                                                      
        flask.g.value = 0                                                      
    return str(flask.g.value)                                                  
                                                                               
with app.app_context() as ctx:                                                 
    with app.test_client() as client:                                          
        resp = client.get(flask.url_for('index'))                              
        app.logger.info(resp)                                                  
        app.logger.info(resp.data)                                             
        # `resp.data` will be '0'                                              
                                                                               
        resp = client.get(flask.url_for('index'))                              
        app.logger.info(resp)                                                  
        app.logger.info(resp.data)                                             
        # `resp.data` will be '1' here ```

**Comment by mitsuhiko** [comment]
Yep. The current behavior is intended behavior indeed. Whenever the last app context closes of the current app it's cleaned up.
